File: code/packages/java/radix-tree/src/main/java/com/codingadventures/radixtree/RadixTree.java
package com.codingadventures.radixtree;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;

public final class RadixTree<V> {
    private final Node<V> root = new Node<>();
    private int size = 0;

    public void insert(String key, V value) {
        if (insertRecursive(root, key, value)) {
            size++;
        }
    }

    public V search(String key) {
        Node<V> node = root;
        String remaining = key;

        while (!remaining.isEmpty()) {
            Edge<V> edge = node.children.get(firstChar(remaining));
            if (edge == null) {
                return null;
            }
            int commonLength = commonPrefixLength(remaining, edge.label);
            if (commonLength < edge.label.length()) {
                return null;
            }
            remaining = remaining.substring(commonLength);
            node = edge.child;
        }

        return node.isEnd ? node.value : null;
    }

    public boolean containsKey(String key) {
        return search(key) != null || keyExists(key);
    }

    public boolean delete(String key) {
        DeleteResult result = deleteRecursive(root, key);
        if (result.deleted) {
            size--;
        }
        return result.deleted;
    }

    public boolean startsWith(String prefix) {
        if (prefix.isEmpty()) {
            return size > 0;
        }

        Node<V> node = root;
        String remaining = prefix;

        while (!remaining.isEmpty()) {
            Edge<V> edge = node.children.get(firstChar(remaining));
            if (edge == null) {
                return false;
            }
            int commonLength = commonPrefixLength(remaining, edge.label);
            if (commonLength == remaining.length()) {
                return true;
            }
            if (commonLength < edge.label.length()) {
                return false;
            }
            remaining = remaining.substring(commonLength);
            node = edge.child;
        }

        return node.isEnd || !node.children.isEmpty();
    }

    public List<String> wordsWithPrefix(String prefix) {
        Node<V> node = root;
        String remaining = prefix;
        StringBuilder path = new StringBuilder();

        if (remaining.isEmpty()) {
            ArrayList<String> results = new ArrayList<>();
            collectKeys(root, "", results);
            return results;
        }

        while (!remaining.isEmpty()) {
            Edge<V> edge = node.children.get(firstChar(remaining));
            if (edge == null) {
                return List.of();
            }
            int commonLength = commonPrefixLength(remaining, edge.label);
            if (commonLength == remaining.length()) {
                if (commonLength == edge.label.length()) {
                    path.append(edge.label);
                    node = edge.child;
                    remaining = "";
                } else {
                    ArrayList<String> results = new ArrayList<>();
                    collectKeys(edge.child, path + edge.label, results);
                    return results;
                }
            } else if (commonLength < edge.label.length()) {
                return List.of();
            } else {
                path.append(edge.label);
                remaining = remaining.substring(commonLength);
                node = edge.child;
            }
        }

        ArrayList<String> results = new ArrayList<>();
        collectKeys(node, path.toString(), results);
        return results;
    }

    public String longestPrefixMatch(String key) {
        Node<V> node = root;
        String remaining = key;
        int consumed = 0;
        String best = node.isEnd ? "" : null;

        while (!remaining.isEmpty()) {
            Edge<V> edge = node.children.get(firstChar(remaining));
            if (edge == null) {
                break;
            }
            int commonLength = commonPrefixLength(remaining, edge.label);
            if (commonLength < edge.label.length()) {
                break;
            }
            consumed += commonLength;
            remaining = remaining.substring(commonLength);
            node = edge.child;
            if (node.isEnd) {
                best = key.substring(0, consumed);
            }
        }

        return best;
    }

    public Map<String, V> toMap() {
        TreeMap<String, V> result = new TreeMap<>();
        collectValues(root, "", result);
        return result;
    }

    public List<String> keys() {
        ArrayList<String> results = new ArrayList<>();
        collectKeys(root, "", results);
        return results;
    }

    public int len() {
        return size;
    }

    public boolean isEmpty() {
        return size == 0;
    }

    public int nodeCount() {
        return countNodes(root);
    }

    @Override
    public String toString() {
        return "RadixTree(" + size + " keys: " + toMap().entrySet().stream().limit(5).toList() + ")";
    }

    private boolean keyExists(String key) {
        Node<V> node = root;
        String remaining = key;
        while (!remaining.isEmpty()) {
            Edge<V> edge = node.children.get(firstChar(remaining));
            if (edge == null) {
                return false;
            }
            int commonLength = commonPrefixLength(remaining, edge.label);
            if (commonLength < edge.label.length()) {
                return false;
            }
            remaining = remaining.substring(commonLength);
            node = edge.child;
        }
        return node.isEnd;
    }

    private static <V> boolean insertRecursive(Node<V> node, String key, V value) {
        if (key.isEmpty()) {
            boolean added = !node.isEnd;
            node.isEnd = true;
            node.value = value;
            return added;
        }

        char first = firstChar(key);
        Edge<V> edge = node.children.remove(first);
        if (edge == null) {
            node.children.put(first, new Edge<>(key, Node.leaf(value)));
            return true;
        }

        int commonLength = commonPrefixLength(key, edge.label);
        if (commonLength == edge.label.length()) {
            boolean added = insertRecursive(edge.child, key.substring(commonLength), value);
            node.children.put(first, edge);
            return added;
        }

        String common = edge.label.substring(0, commonLength);
        String labelRest = edge.label.substring(commonLength);
        String keyRest = key.substring(commonLength);
        Node<V> splitNode = new Node<>();
        splitNode.children.put(firstChar(labelRest), new Edge<>(labelRest, edge.child));

        if (keyRest.isEmpty()) {
            splitNode.isEnd = true;
            splitNode.value = value;
        } else {
            splitNode.children.put(firstChar(keyRest), new Edge<>(keyRest, Node.leaf(value)));
        }

        node.children.put(firstChar(common), new Edge<>(common, splitNode));
        return true;
    }

    private static <V> DeleteResult deleteRecursive(Node<V> node, String key) {
        if (key.isEmpty()) {
            if (!node.isEnd) {
                return new DeleteResult(false, false);
            }
            node.isEnd = false;
            node.value = null;
            return new DeleteResult(true, !node.isEnd && node.children.size() == 1);
        }

        char first = firstChar(key);
        Edge<V> edge = node.children.remove(first);
        if (edge == null) {
            return new DeleteResult(false, false);
        }

        int commonLength = commonPrefixLength(key, edge.label);
        if (commonLength < edge.label.length()) {
            node.children.put(first, edge);
            return new DeleteResult(false, false);
        }

        DeleteResult result = deleteRecursive(edge.child, key.substring(commonLength));
        if (!result.deleted) {
            node.children.put(first, edge);
            return result;
        }

        if (result.childMergeable) {
            Map.Entry<Character, Edge<V>> grandchildEntry = edge.child.children.firstEntry();
            Edge<V> grandchild = grandchildEntry.getValue();
            String mergedLabel = edge.label + grandchild.label;
            node.children.put(firstChar(mergedLabel), new Edge<>(mergedLabel, grandchild.child));
        } else if (!edge.child.isEnd && edge.child.children.isEmpty()) {
            // Drop the empty child edge.
        } else {
            node.children.put(first, edge);
        }

        return new DeleteResult(true, !node.isEnd && node.children.size() == 1);
    }

    private static <V> void collectKeys(Node<V> node, String current, List<String> results) {
        if (node.isEnd) {
            results.add(current);
        }
        for (Edge<V> edge : node.children.values()) {
            collectKeys(edge.child, current + edge.label, results);
        }
    }

    private static <V> void collectValues(Node<V> node, String current, Map<String, V> result) {
        if (node.isEnd) {
            result.put(current, node.value);
        }
        for (Edge<V> edge : node.children.values()) {
            collectValues(edge.child, current + edge.label, result);
        }
    }

    private static <V> int countNodes(Node<V> node) {
        int count = 1;
        for (Edge<V> edge : node.children.values()) {
            count += countNodes(edge.child);
        }
        return count;
    }

    private static int commonPrefixLength(String left, String right) {
        int index = 0;
        int limit = Math.min(left.length(), right.length());
        while (index < limit && left.charAt(index) == right.charAt(index)) {
            index++;
        }
        return index;
    }

    private static char firstChar(String value) {
        return value.charAt(0);
    }

    private static final class Node<V> {
        private boolean isEnd;
        private V value;
        private final TreeMap<Character, Edge<V>> children = new TreeMap<>();

        private static <V> Node<V> leaf(V value) {
            Node<V> node = new Node<>();
            node.isEnd = true;
            node.value = value;
            return node;
        }
    }

    private record Edge<V>(String label, Node<V> child) {
    }

    private record DeleteResult(boolean deleted, boolean childMergeable) {
    }
}

File: code/packages/kotlin/radix-tree/src/main/kotlin/com/codingadventures/radixtree/RadixTree.kt
package com.codingadventures.radixtree

import java.util.TreeMap

class RadixTree<V> {
    private val root = Node<V>()
    private var size = 0

    fun insert(key: String, value: V) {
        if (insertRecursive(root, key, value)) size += 1
    }

    fun search(key: String): V? {
        var node = root
        var remaining = key
        while (remaining.isNotEmpty()) {
            val edge = node.children[firstChar(remaining)] ?: return null
            val commonLength = commonPrefixLength(remaining, edge.label)
            if (commonLength < edge.label.length) return null
            remaining = remaining.substring(commonLength)
            node = edge.child
        }
        return if (node.isEnd) node.value else null
    }

    fun containsKey(key: String): Boolean = search(key) != null || keyExists(key)

    fun delete(key: String): Boolean {
        val result = deleteRecursive(root, key)
        if (result.deleted) size -= 1
        return result.deleted
    }

    fun startsWith(prefix: String): Boolean {
        if (prefix.isEmpty()) return size > 0
        var node = root
        var remaining = prefix
        while (remaining.isNotEmpty()) {
            val edge = node.children[firstChar(remaining)] ?: return false
            val commonLength = commonPrefixLength(remaining, edge.label)
            if (commonLength == remaining.length) return true
            if (commonLength < edge.label.length) return false
            remaining = remaining.substring(commonLength)
            node = edge.child
        }
        return node.isEnd || node.children.isNotEmpty()
    }

    fun wordsWithPrefix(prefix: String): List<String> {
        var node = root
        var remaining = prefix
        val path = StringBuilder()

        if (remaining.isEmpty()) {
            return buildList { collectKeys(root, "", this) }
        }

        while (remaining.isNotEmpty()) {
            val edge = node.children[firstChar(remaining)] ?: return emptyList()
            val commonLength = commonPrefixLength(remaining, edge.label)
            if (commonLength == remaining.length) {
                if (commonLength == edge.label.length) {
                    path.append(edge.label)
                    node = edge.child
                    remaining = ""
                } else {
                    return buildList { collectKeys(edge.child, path.toString() + edge.label, this) }
                }
            } else if (commonLength < edge.label.length) {
                return emptyList()
            } else {
                path.append(edge.label)
                remaining = remaining.substring(commonLength)
                node = edge.child
            }
        }

        return buildList { collectKeys(node, path.toString(), this) }
    }

    fun longestPrefixMatch(key: String): String? {
        var node = root
        var remaining = key
        var consumed = 0
        var best: String? = if (node.isEnd) "" else null

        while (remaining.isNotEmpty()) {
            val edge = node.children[firstChar(remaining)] ?: break
            val commonLength = commonPrefixLength(remaining, edge.label)
            if (commonLength < edge.label.length) break
            consumed += commonLength
            remaining = remaining.substring(commonLength)
            node = edge.child
            if (node.isEnd) best = key.substring(0, consumed)
        }

        return best
    }

    fun toMap(): Map<String, V> = sortedMapOf<String, V>().also { collectValues(root, "", it) }
    fun keys(): List<String> = buildList { collectKeys(root, "", this) }
    fun len(): Int = size
    fun isEmpty(): Boolean = size == 0
    fun nodeCount(): Int = countNodes(root)

    override fun toString(): String = "RadixTree($size keys: ${toMap().entries.take(5)})"

    private fun keyExists(key: String): Boolean {
        var node = root
        var remaining = key
        while (remaining.isNotEmpty()) {
            val edge = node.children[firstChar(remaining)] ?: return false
            val commonLength = commonPrefixLength(remaining, edge.label)
            if (commonLength < edge.label.length) return false
            remaining = remaining.substring(commonLength)
            node = edge.child
        }
        return node.isEnd
    }

    private fun insertRecursive(node: Node<V>, key: String, value: V): Boolean {
        if (key.isEmpty()) {
            val added = !node.isEnd
            node.isEnd = true
            node.value = value
            return added
        }

        val first = firstChar(key)
        val edge = node.children.remove(first)
        if (edge == null) {
            node.children[first] = Edge(key, Node.leaf(value))
            return true
        }

        val commonLength = commonPrefixLength(key, edge.label)
        if (commonLength == edge.label.length) {
            val added = insertRecursive(edge.child, key.substring(commonLength), value)
            node.children[first] = edge
            return added
        }

        val common = edge.label.substring(0, commonLength)
        val labelRest = edge.label.substring(commonLength)
        val keyRest = key.substring(commonLength)
        val splitNode = Node<V>()
        splitNode.children[firstChar(labelRest)] = Edge(labelRest, edge.child)

        if (keyRest.isEmpty()) {
            splitNode.isEnd = true
            splitNode.value = value
        } else {
            splitNode.children[firstChar(keyRest)] = Edge(keyRest, Node.leaf(value))
        }

        node.children[firstChar(common)] = Edge(common, splitNode)
        return true
    }

    private fun deleteRecursive(node: Node<V>, key: String): DeleteResult {
        if (key.isEmpty()) {
            if (!node.isEnd) return DeleteResult(false, false)
            node.isEnd = false
            node.value = null
            return DeleteResult(true, !node.isEnd && node.children.size == 1)
        }

        val first = firstChar(key)
        val edge = node.children.remove(first) ?: return DeleteResult(false, false)
        val commonLength = commonPrefixLength(key, edge.label)
        if (commonLength < edge.label.length) {
            node.children[first] = edge
            return DeleteResult(false, false)
        }

        val result = deleteRecursive(edge.child, key.substring(commonLength))
        if (!result.deleted) {
            node.children[first] = edge
            return result
        }

        when {
            result.childMergeable -> {
                val grandchild = edge.child.children.firstEntry().value
                val mergedLabel = edge.label + grandchild.label
                node.children[firstChar(mergedLabel)] = Edge(mergedLabel, grandchild.child)
            }
            !edge.child.isEnd && edge.child.children.isEmpty() -> Unit
            else -> node.children[first] = edge
        }

        return DeleteResult(true, !node.isEnd && node.children.size == 1)
    }

    private fun collectKeys(node: Node<V>, current: String, results: MutableList<String>) {
        if (node.isEnd) results += current
        node.children.values.forEach { edge -> collectKeys(edge.child, current + edge.label, results) }
    }

    private fun collectValues(node: Node<V>, current: String, results: MutableMap<String, V>) {
        if (node.isEnd) results[current] = node.value as V
        node.children.values.forEach { edge -> collectValues(edge.child, current + edge.label, results) }
    }

    private fun countNodes(node: Node<V>): Int = 1 + node.children.values.sumOf { countNodes(it.child) }

    private fun commonPrefixLength(left: String, right: String): Int {
        var index = 0
        val limit = minOf(left.length, right.length)
        while (index < limit && left[index] == right[index]) index += 1
        return index
    }

    private fun firstChar(value: String): Char = value[0]

    private class Node<V> {
        var isEnd = false
        var value: V? = null
        val children = TreeMap<Char, Edge<V>>()

        companion object {
            fun <V> leaf(value: V): Node<V> = Node<V>().also {
                it.isEnd = true
                it.value = value
            }
        }
    }

    private data class Edge<V>(val label: String, val child: Node<V>)
    private data class DeleteResult(val deleted: Boolean, val childMergeable: Boolean)
}
